File: packages/jetbrains-plugin/src/main/java/com/chainlesschain/ide/PluginQuality.java
package com.chainlesschain.ide;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Plugin / LSP quality board (gap #11) — the Java twin of the VS Code
 * extension's {@code plugin-quality.js}. Combines three CLI surfaces:
 *
 * <pre>
 *   cc plugin installed --json        → [{name, version, scope, dir, ok}]
 *   cc plugin validate &lt;dir&gt; --json   → {ok, errors, warnings, components,
 *                                        componentCounts:{skills,agents,hooks,
 *                                        mcp,lsp,monitors,bin,settings}, …}
 *   cc code-intel status --json       → {servers:[{languageId,id,available,
 *                                        command}], extensions:[…]}
 * </pre>
 *
 * into per-plugin quality rows: contributed component counts by type + flags:
 * <ul>
 *   <li>{@code broken} — validate reported errors / invalid manifest;</li>
 *   <li>{@code lsp} — {@code none | ok | unavailable | unknown}, derived from
 *       the live {@code code-intel status} probe; when the probe result can't
 *       be matched to the plugin's own declared server (plugin untrusted → not
 *       registered, or status unreadable) we say "unknown" instead of
 *       fabricating a verdict;</li>
 *   <li>{@code unused} — contributes NOTHING (every component count is zero).
 *       A plugin that only ships an LSP server is NOT unused.</li>
 * </ul>
 *
 * There is deliberately NO {@code slow} flag: the CLI plugin runtime records
 * no load/exec timing today, and we don't fake metrics we don't have.
 *
 * Pure JDK (no IntelliJ SDK, no process spawning) → JUnit + smoke tested; the
 * Swing glue (the "Quality" tab of the Plugins &amp; MCP manager) lives in
 * {@code intellij/PluginManagerAction}. Output is plain text for a monospace
 * area — no HTML, so no escaping is needed.
 */
public final class PluginQuality {

    private PluginQuality() {}

    /** Component types {@code plugin validate --json} counts (manifest summarizeComponents). */
    public static final List<String> COMPONENT_TYPES = List.of(
            "skills", "agents", "hooks", "mcp", "lsp", "monitors", "bin", "settings");

    // LSP availability verdicts (deriveLspAvailability).
    public static final String LSP_NONE = "none";
    public static final String LSP_OK = "ok";
    public static final String LSP_UNAVAILABLE = "unavailable";
    public static final String LSP_UNKNOWN = "unknown";

    // ------------------------------------------------------------- cc args

    public static List<String> buildPluginValidateArgs(String dir) {
        return Arrays.asList("plugin", "validate", String.valueOf(dir), "--json");
    }

    public static List<String> buildCodeIntelStatusArgs() {
        return Arrays.asList("code-intel", "status", "--json");
    }

    // --------------------------------------------------------------- model

    /** An LSP server a plugin declares (from validate's {@code components.lsp}). */
    public static final class LspEntry {
        public final String languageId;
        public final String id;

        public LspEntry(String languageId, String id) {
            this.languageId = nz(languageId);
            this.id = nz(id);
        }
    }

    /**
     * Parsed {@code plugin validate} output for one plugin, or a per-plugin
     * failure marker ({@link #failure}) when the CLI produced no usable JSON —
     * a failed validate NEVER fabricates flags, it degrades to "unknown".
     */
    public static final class Validation {
        public final boolean failed;
        public final String message;
        public final boolean ok;
        public final List<String> errors;
        public final List<String> warnings;
        /** Per-type component counts over {@link #COMPONENT_TYPES}; null when failed. */
        public final Map<String, Integer> counts;
        public final List<LspEntry> lsp;

        private Validation(boolean failed, String message, boolean ok,
                List<String> errors, List<String> warnings,
                Map<String, Integer> counts, List<LspEntry> lsp) {
            this.failed = failed;
            this.message = nz(message);
            this.ok = ok;
            this.errors = errors == null ? List.of() : errors;
            this.warnings = warnings == null ? List.of() : warnings;
            this.counts = counts;
            this.lsp = lsp == null ? List.of() : lsp;
        }

        /** Per-plugin tolerance marker: validate ran but yielded nothing usable. */
        public static Validation failure(String message) {
            return new Validation(true, message, false, null, null, null, null);
        }
    }

    /** One row of the live {@code code-intel status} probe. */
    public static final class StatusServer {
        public final String languageId;
        public final String id;
        public final boolean available;

        public StatusServer(String languageId, String id, boolean available) {
            this.languageId = nz(languageId);
            this.id = nz(id);
            this.available = available;
        }
    }

    // --------------------------------------------------------------- parse

    /**
     * {@code plugin validate <dir> --json} → {@link Validation}; null =
     * unreadable. The command prints the JSON even for an INVALID manifest (it
     * only sets the exit code), so null really means "no usable output".
     */
    public static Validation parsePluginValidate(String text) {
        Map<?, ?> parsed = parseObject(text);
        if (parsed == null) return null;
        Object rawCounts = parsed.get("componentCounts");
        Map<?, ?> countsMap = rawCounts instanceof Map ? (Map<?, ?>) rawCounts : Map.of();
        Map<String, Integer> counts = new LinkedHashMap<String, Integer>();
        for (String t : COMPONENT_TYPES) {
            Object v = countsMap.get(t);
            int n = v instanceof Number ? ((Number) v).intValue() : 0;
            counts.put(t, n > 0 ? n : 0);
        }
        Object components = parsed.get("components");
        Object lspRaw = components instanceof Map ? ((Map<?, ?>) components).get("lsp") : null;
        List<LspEntry> lsp = new ArrayList<LspEntry>();
        if (lspRaw instanceof List) {
            for (Object o : (List<?>) lspRaw) {
                if (!(o instanceof Map)) continue;
                Map<?, ?> m = (Map<?, ?>) o;
                if (!(m.get("languageId") instanceof String)) continue;
                String id = str(m.get("id"));
                if (id.isEmpty()) id = str(m.get("command"));
                lsp.add(new LspEntry((String) m.get("languageId"), id));
            }
        }
        return new Validation(false, "", Boolean.TRUE.equals(parsed.get("ok")),
                strList(parsed.get("errors")), strList(parsed.get("warnings")),
                counts, lsp);
    }

    /** {@code code-intel status --json} → server rows; null = unreadable. */
    public static List<StatusServer> parseCodeIntelStatus(String text) {
        Map<?, ?> parsed = parseObject(text);
        if (parsed == null || !(parsed.get("servers") instanceof List)) return null;
        List<StatusServer> out = new ArrayList<StatusServer>();
        for (Object o : (List<?>) parsed.get("servers")) {
            if (!(o instanceof Map)) continue;
            Map<?, ?> m = (Map<?, ?>) o;
            if (!(m.get("languageId") instanceof String)) continue;
            out.add(new StatusServer((String) m.get("languageId"),
                    str(m.get("id")), Boolean.TRUE.equals(m.get("available"))));
        }
        return out;
    }

    // ------------------------------------------------------------- derive

    /**
     * Derive a plugin's LSP availability from the live probe.
     *
     * <ul>
     *   <li>{@code none} — the plugin declares no LSP servers;</li>
     *   <li>{@code unknown} — no live status data, OR the status row for that
     *       languageId belongs to a DIFFERENT server (the plugin is
     *       untrusted/not registered, so its own binary was never probed) —
     *       we do not fabricate a verdict from someone else's probe;</li>
     *   <li>{@code unavailable} — the plugin's own registered server did not
     *       resolve a binary;</li>
     *   <li>{@code ok} — every declared server resolved.</li>
     * </ul>
     */
    public static String deriveLspAvailability(List<LspEntry> lspEntries,
            List<StatusServer> statusServers) {
        List<LspEntry> entries = lspEntries == null ? List.of() : lspEntries;
        if (entries.isEmpty()) return LSP_NONE;
        if (statusServers == null) return LSP_UNKNOWN;
        boolean sawUnknown = false;
        for (LspEntry e : entries) {
            StatusServer row = null;
            for (StatusServer s : statusServers) {
                if (s.languageId.equals(e.languageId)) { row = s; break; }
            }
            if (row == null || (!e.id.isEmpty() && !row.id.equals(e.id))) {
                sawUnknown = true;
                continue;
            }
            if (!row.available) return LSP_UNAVAILABLE;
        }
        return sawUnknown ? LSP_UNKNOWN : LSP_OK;
    }

    // ---------------------------------------------------------------- rows

    /**
     * One quality row. Tri-state {@code broken}/{@code unused} ({@code
     * Boolean.TRUE}/{@code FALSE}/null): null means "validate output
     * unavailable — unknown", never a fabricated verdict. There is no
     * {@code slow} field on purpose (no timing data exists).
     */
    public static final class Row {
        public final String name;
        public final String version;
        public final String scope;
        /** Component counts; null when validate output was unavailable. */
        public final Map<String, Integer> counts;
        public final Boolean broken;
        public final List<String> brokenReasons;
        /** {@link #LSP_NONE} | {@link #LSP_OK} | {@link #LSP_UNAVAILABLE} | {@link #LSP_UNKNOWN}. */
        public final String lsp;
        public final Boolean unused;
        public final boolean validateFailed;
        public final String validateMessage;

        Row(String name, String version, String scope, Map<String, Integer> counts,
                Boolean broken, List<String> brokenReasons, String lsp,
                Boolean unused, boolean validateFailed, String validateMessage) {
            this.name = nz(name);
            this.version = nz(version);
            this.scope = scope == null || scope.isEmpty() ? "user" : scope;
            this.counts = counts;
            this.broken = broken;
            this.brokenReasons = brokenReasons == null ? List.of() : brokenReasons;
            this.lsp = nz(lsp);
            this.unused = unused;
            this.validateFailed = validateFailed;
            this.validateMessage = nz(validateMessage);
        }
    }

    /**
     * Build quality rows from the three sources.
     *
     * @param plugins     rows from {@code PluginManager.parsePluginInstalled}
     *                    (name/version/scope/dir/ok maps); null = unreadable
     * @param validations per-plugin {@link Validation} by name (a missing or
     *                    failed entry degrades that row honestly)
     * @param lspStatus   {@link #parseCodeIntelStatus} result; null = no probe
     * @return rows, or null when the plugin list itself was unreadable.
     */
    public static List<Row> buildQualityRows(List<Map<String, Object>> plugins,
            Map<String, Validation> validations, List<StatusServer> lspStatus) {
        if (plugins == null) return null;
        List<Row> out = new ArrayList<Row>();
        for (Map<String, Object> p : plugins) {
            if (p == null) continue;
            String name = str(p.get("name"));
            String version = str(p.get("version"));
            String scope = str(p.get("scope"));
            boolean discoveredBad = Boolean.FALSE.equals(p.get("ok"));
            Validation v = validations == null ? null : validations.get(name);
            boolean failed = v == null || v.failed || v.counts == null;
            if (failed) {
                out.add(new Row(name, version, scope, null,
                        discoveredBad ? Boolean.TRUE : null, List.of(),
                        LSP_UNKNOWN, null, true,
                        v != null && !v.message.isEmpty()
                                ? v.message : "plugin validate produced no output"));
                continue;
            }
            boolean broken = !v.ok || !v.errors.isEmpty() || discoveredBad;
            boolean unused = true;
            for (String t : COMPONENT_TYPES) {
                Integer n = v.counts.get(t);
                if (n != null && n > 0) { unused = false; break; }
            }
            out.add(new Row(name, version, scope,
                    Collections.unmodifiableMap(v.counts), broken,
                    v.errors.subList(0, Math.min(5, v.errors.size())),
                    deriveLspAvailability(v.lsp, lspStatus), unused, false, ""));
        }
        return out;
    }

    // ------------------------------------------------------------- display

    /** {@code "skills 2 · lsp 1"} — only the non-zero component types. */
    public static String formatCounts(Map<String, Integer> counts) {
        if (counts == null) return "";
        StringBuilder sb = new StringBuilder();
        for (String t : COMPONENT_TYPES) {
            Integer n = counts.get(t);
            if (n == null || n <= 0) continue;
            if (sb.length() > 0) sb.append(" · ");
            sb.append(t).append(' ').append(n);
        }
        return sb.toString();
    }

    /**
     * The flag badges for one row (plain text, same rules as the VS twin's
     * pills): {@code ✖ broken} / {@code validity unknown}, the lsp verdict
     * (silent when {@code none}), and {@code unused}. Deliberately no
     * {@code slow} badge — the CLI has no timing.
     */
    public static List<String> flagsFor(Row r) {
        List<String> flags = new ArrayList<String>();
        if (r == null) return flags;
        if (Boolean.TRUE.equals(r.broken)) flags.add("✖ broken");
        else if (r.broken == null) flags.add("validity unknown");
        if (LSP_OK.equals(r.lsp)) flags.add("lsp ok");
        else if (LSP_UNAVAILABLE.equals(r.lsp)) flags.add("lsp unavailable");
        else if (LSP_UNKNOWN.equals(r.lsp)) flags.add("lsp unknown");
        if (Boolean.TRUE.equals(r.unused)) flags.add("unused");
        return flags;
    }

    /**
     * Plain-text quality board for a monospace area (the JB stand-in for the
     * VS twin's {@code renderQualityHtml}): probe warning, unreadable/empty
     * states, then one block per plugin — name/version/scope, component
     * counts (or the per-plugin validate failure), flags and broken reasons.
     */
    public static String describe(List<Row> rows, boolean lspStatusAvailable) {
        StringBuilder sb = new StringBuilder();
        if (!lspStatusAvailable) {
            sb.append("⚠ `cc code-intel status` unavailable — LSP availability shown as unknown.\n\n");
        }
        if (rows == null) {
            sb.append("could not read plugins (is cc installed?)");
            return sb.toString();
        }
        if (rows.isEmpty()) {
            sb.append("No runtime plugins installed.");
            return sb.toString();
        }
        boolean first = true;
        for (Row r : rows) {
            if (r == null) continue;
            if (!first) sb.append('\n');
            first = false;
            sb.append(r.name);
            if (!r.version.isEmpty()) sb.append(" v").append(r.version);
            sb.append("  [").append(r.scope).append("]\n");
            if (r.validateFailed) {
                sb.append("    components: validate failed: ")
                        .append(r.validateMessage).append('\n');
            } else {
                String counts = formatCounts(r.counts);
                sb.append("    components: ")
                        .append(counts.isEmpty() ? "(none)" : counts).append('\n');
            }
            List<String> flags = flagsFor(r);
            if (!flags.isEmpty()) {
                sb.append("    flags: ").append(String.join(", ", flags)).append('\n');
            }
            if (!r.brokenReasons.isEmpty()) {
                sb.append("    errors: ")
                        .append(String.join("; ", r.brokenReasons)).append('\n');
            }
        }
        return sb.toString();
    }

    /**
     * One-line counts summary for the board header:
     * {@code "plugins: 3 · broken: 1 · lsp unavailable: 1 · unknown: 1 · unused: 1"}
     * (plus the probe warning). {@code "plugins: n/a"} when unreadable.
     */
    public static String summaryLine(List<Row> rows, boolean lspStatusAvailable) {
        if (rows == null) return "plugins: n/a";
        int broken = 0, lspUnavailable = 0, unknown = 0, unused = 0;
        for (Row r : rows) {
            if (r == null) continue;
            if (Boolean.TRUE.equals(r.broken)) broken++;
            if (LSP_UNAVAILABLE.equals(r.lsp)) lspUnavailable++;
            if (r.broken == null || LSP_UNKNOWN.equals(r.lsp)) unknown++;
            if (Boolean.TRUE.equals(r.unused)) unused++;
        }
        StringBuilder sb = new StringBuilder();
        sb.append("plugins: ").append(rows.size())
                .append(" · broken: ").append(broken)
                .append(" · lsp unavailable: ").append(lspUnavailable)
                .append(" · unknown: ").append(unknown)
                .append(" · unused: ").append(unused);
        if (!lspStatusAvailable) sb.append(" · lsp probe: unavailable");
        return sb.toString();
    }

    // -------------------------------------------------------------- helpers

    private static Map<?, ?> parseObject(String text) {
        try {
            Object parsed = MiniJson.parse(text == null ? "" : text.trim());
            return parsed instanceof Map ? (Map<?, ?>) parsed : null;
        } catch (RuntimeException e) {
            return null;
        }
    }

    private static List<String> strList(Object v) {
        List<String> out = new ArrayList<String>();
        if (v instanceof List) {
            for (Object o : (List<?>) v) {
                if (o == null) continue;
                String s = String.valueOf(o);
                if (!s.isEmpty()) out.add(s);
            }
        }
        return out;
    }

    private static String nz(String s) {
        return s == null ? "" : s;
    }

    private static String str(Object v) {
        return v == null ? "" : String.valueOf(v);
    }
}

File: packages/vscode-extension/src/plugin-quality.js
/**
 * Plugin / LSP quality board (gap #11) — pure logic for the "Quality" section
 * of the Plugins & MCP manager panel. Combines three CLI surfaces:
 *
 *   cc plugin installed --json        → [{name, version, scope, dir, ok}]
 *   cc plugin validate <dir> --json   → {ok, errors, warnings, components,
 *                                        componentCounts:{skills,agents,hooks,
 *                                        mcp,lsp,monitors,bin,settings}, …}
 *   cc code-intel status --json       → {servers:[{languageId,id,available,
 *                                        command}], extensions:[…]}
 *
 * into per-plugin quality rows: contributed component counts by type + flags:
 *   broken          validate reported errors / invalid manifest
 *   lsp             'none' | 'ok' | 'unavailable' | 'unknown' — derived from
 *                   the live `code-intel status` probe; when the probe result
 *                   can't be matched to the plugin's declared server (plugin
 *                   untrusted → not registered, or status unreadable) we say
 *                   "unknown" instead of fabricating a verdict.
 *   unused          contributes NOTHING (every component count is zero). A
 *                   plugin that only ships an LSP server is NOT unused.
 *
 * There is deliberately NO `slow` flag: the CLI plugin runtime records no
 * load/exec timing today, and we don't fake metrics we don't have.
 *
 * Pure Node (no `vscode`) → unit-testable; the webview glue lives in
 * ui/plugin-manager-view.js. All strings that reach HTML go through
 * escapeHtml — names/errors come from user-authored manifests.
 */

/** Component types `plugin validate --json` counts (manifest.js summarizeComponents). */
const COMPONENT_TYPES = [
  "skills",
  "agents",
  "hooks",
  "mcp",
  "lsp",
  "monitors",
  "bin",
  "settings",
];

function buildPluginValidateArgs(dir) {
  return ["plugin", "validate", String(dir), "--json"];
}

function buildCodeIntelStatusArgs() {
  return ["code-intel", "status", "--json"];
}

function parseObject(text) {
  try {
    const parsed = JSON.parse(String(text || "").trim());
    return parsed && typeof parsed === "object" && !Array.isArray(parsed)
      ? parsed
      : null;
  } catch {
    return null;
  }
}

/**
 * `plugin validate <dir> --json` → {ok, errors, warnings, counts, lsp};
 * null = unreadable. The command prints the JSON even for an INVALID manifest
 * (it only sets the exit code), so null really means "no usable output".
 */
function parsePluginValidate(text) {
  const parsed = parseObject(text);
  if (!parsed) return null;
  const rawCounts =
    parsed.componentCounts && typeof parsed.componentCounts === "object"
      ? parsed.componentCounts
      : {};
  const counts = {};
  for (const t of COMPONENT_TYPES) {
    const n = Number(rawCounts[t]);
    counts[t] = Number.isFinite(n) && n > 0 ? n : 0;
  }
  const strList = (v) =>
    Array.isArray(v) ? v.map((e) => String(e)).filter(Boolean) : [];
  const lspRaw = parsed.components && parsed.components.lsp;
  const lsp = (Array.isArray(lspRaw) ? lspRaw : [])
    .filter((s) => s && typeof s.languageId === "string")
    .map((s) => ({
      languageId: s.languageId,
      id: String(s.id || s.command || ""),
    }));
  return {
    ok: parsed.ok === true,
    errors: strList(parsed.errors),
    warnings: strList(parsed.warnings),
    counts,
    lsp,
  };
}

/** `code-intel status --json` → [{languageId, id, available}]; null = unreadable. */
function parseCodeIntelStatus(text) {
  const parsed = parseObject(text);
  if (!parsed || !Array.isArray(parsed.servers)) return null;
  return parsed.servers
    .filter((s) => s && typeof s.languageId === "string")
    .map((s) => ({
      languageId: s.languageId,
      id: String(s.id || ""),
      available: s.available === true,
    }));
}

/**
 * Derive a plugin's LSP availability from the live `code-intel status` probe.
 *
 *  'none'        the plugin declares no LSP servers
 *  'unknown'     no live status data, OR the status row for that languageId
 *                belongs to a DIFFERENT server (the plugin is untrusted/not
 *                registered, so its own binary was never probed) — we do not
 *                fabricate a verdict from someone else's probe
 *  'unavailable' the plugin's own registered server did not resolve a binary
 *  'ok'          every declared server resolved
 */
function deriveLspAvailability(lspEntries, statusServers) {
  const entries = Array.isArray(lspEntries) ? lspEntries : [];
  if (!entries.length) return "none";
  if (!Array.isArray(statusServers)) return "unknown";
  let sawUnknown = false;
  for (const e of entries) {
    const row = statusServers.find((s) => s.languageId === e.languageId);
    if (!row || (e.id && row.id !== e.id)) {
      sawUnknown = true;
      continue;
    }
    if (!row.available) return "unavailable";
  }
  return sawUnknown ? "unknown" : "ok";
}

/**
 * Build quality rows from the three sources.
 *
 * @param {{
 *   plugins: Array<{name,version,scope,dir,ok}>|null,  // parsePluginInstalled
 *   validations?: Object<string, {failed?:boolean,message?:string}|
 *                                ReturnType<parsePluginValidate>>, // by name
 *   lspStatus?: Array|null,                            // parseCodeIntelStatus
 * }} sources
 * @returns {Array|null} rows; null when the plugin list itself was unreadable.
 *
 * Row: {name, version, scope, counts|null, broken:true|false|null,
 *       brokenReasons:string[], lsp:'none'|'ok'|'unavailable'|'unknown',
 *       unused:true|false|null, validateFailed, validateMessage}
 * `null` flag values mean "validate output unavailable — unknown", never a
 * fabricated verdict.
 */
function buildQualityRows({ plugins, validations = {}, lspStatus = null } = {}) {
  if (!Array.isArray(plugins)) return null;
  return plugins.map((p) => {
    const v = validations ? validations[p.name] : null;
    const failed = !v || v.failed === true || !v.counts;
    if (failed) {
      return {
        name: p.name,
        version: p.version || "",
        scope: p.scope || "user",
        counts: null,
        broken: p.ok === false ? true : null,
        brokenReasons: [],
        lsp: "unknown",
        unused: null,
        validateFailed: true,
        validateMessage:
          (v && v.message) || "plugin validate produced no output",
      };
    }
    const broken = v.ok === false || v.errors.length > 0 || p.ok === false;
    return {
      name: p.name,
      version: p.version || "",
      scope: p.scope || "user",
      counts: v.counts,
      broken,
      brokenReasons: v.errors.slice(0, 5),
      lsp: deriveLspAvailability(v.lsp, lspStatus),
      unused: COMPONENT_TYPES.every((t) => !v.counts[t]),
      validateFailed: false,
      validateMessage: "",
    };
  });
}

function escapeHtml(value) {
  return String(value == null ? "" : value).replace(
    /[&<>"']/g,
    (c) =>
      ({
        "&": "&amp;",
        "<": "&lt;",
        ">": "&gt;",
        '"': "&quot;",
        "'": "&#39;",
      })[c],
  );
}

/** "skills 2 · lsp 1" — only the non-zero component types. */
function formatCounts(counts) {
  if (!counts) return "";
  return COMPONENT_TYPES.filter((t) => counts[t] > 0)
    .map((t) => `${t} ${counts[t]}`)
    .join(" · ");
}

/**
 * Quality-section HTML fragment (server-rendered, fully escaped) for the
 * plugin manager webview. Reuses the panel's existing CSS classes
 * (ok/bad/muted/pill/err/warnflag).
 */
function renderQualityHtml(rows, { lspStatusAvailable = true } = {}) {
  const parts = [];
  if (!lspStatusAvailable) {
    parts.push(
      '<p class="warnflag">⚠ `cc code-intel status` unavailable — LSP availability shown as unknown.</p>',
    );
  }
  if (rows === null) {
    parts.push(
      '<p class="err">could not read plugins (is cc installed?)</p>',
    );
    return parts.join("");
  }
  if (!rows.length) {
    parts.push('<p class="muted">No runtime plugins installed.</p>');
    return parts.join("");
  }
  const lspBadge = {
    ok: '<span class="pill ok">lsp ok</span>',
    unavailable: '<span class="pill bad">lsp unavailable</span>',
    unknown: '<span class="pill muted">lsp unknown</span>',
    none: "",
  };
  const body = rows
    .map((r) => {
      const flags = [];
      if (r.broken === true) {
        flags.push(
          `<span class="pill bad" title="${escapeHtml(r.brokenReasons.join("; "))}">✖ broken</span>`,
        );
      } else if (r.broken === null) {
        flags.push('<span class="pill muted">validity unknown</span>');
      }
      flags.push(lspBadge[r.lsp] || "");
      if (r.unused === true) {
        flags.push(
          '<span class="pill warnflag" title="contributes no skills/agents/hooks/mcp/monitors/bin/settings and no lsp">unused</span>',
        );
      }
      const components = r.validateFailed
        ? `<span class="muted">validate failed: ${escapeHtml(r.validateMessage)}</span>`
        : escapeHtml(formatCounts(r.counts)) ||
          '<span class="muted">(none)</span>';
      return (
        `<tr><td>${escapeHtml(r.name)} <span class="muted">v${escapeHtml(r.version)}</span></td>` +
        `<td><span class="pill">${escapeHtml(r.scope)}</span></td>` +
        `<td>${components}</td>` +
        `<td>${flags.filter(Boolean).join(" ")}</td></tr>`
      );
    })
    .join("");
  parts.push(
    '<table><thead><tr><th>plugin</th><th>scope</th><th>components</th><th style="width:260px">flags</th></tr></thead><tbody>' +
      body +
      "</tbody></table>",
  );
  return parts.join("");
}

module.exports = {
  COMPONENT_TYPES,
  buildPluginValidateArgs,
  buildCodeIntelStatusArgs,
  parsePluginValidate,
  parseCodeIntelStatus,
  deriveLspAvailability,
  buildQualityRows,
  formatCounts,
  escapeHtml,
  renderQualityHtml,
};
